IBX-340: Allowed URL Gateway queries to not perform count operations

Steveb-p opened the pull request:
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-340](https://issues.ibexa.co/browse/IBX-340)
| **Type**                                   | improvement
| **Target eZ Platform version** | `v3.3`
| **BC breaks**                          | no
| **Doc needed**                       | no

Prevent multiple database `count` queries when paginating URL lists.

Without the patch:
![image](https://user-images.githubusercontent.com/3183926/117416703-eff84300-af19-11eb-8d51-01b7f8d38671.png)


With the patch:
![image](https://user-images.githubusercontent.com/3183926/117416664-e1aa2700-af19-11eb-842a-9f7633ccc8f8.png)


#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/php-dev-team`).

lserwatka commented:
@Steveb-p we need a JIRA ticket for this PR, so our automatic changelog generator can hook it.

Steveb-p commented:
@lserwatka a proper Jira ticket has been added: [IBX-340](https://issues.ibexa.co/browse/IBX-340)

adamwojs reviewed (commented):
+1 but please create/update integration test. It's   PAPI change

Steveb-p commented:
I've added some additional changes that I wanted to make a different PR with, but considering they are related to the changes here I'm including them now.

1. `eZ\Publish\Core\FieldType\Url\UrlStorage\Gateway\DoctrineStorage` now shares constants with `eZ\Publish\Core\Persistence\Legacy\URL\Gateway\DoctrineDatabase` (for discoverability and to prevent making changes to one but not the other).
2. Tests do not make use of constants (to detect changes to those)
3. I've added coverage annotation to test cases I've touched. I'd like to advocate usage of class-level coverage annotation instead of method based one, as they are not endorsed by phpunit. They are also not properly detected by phpstorm afaik; for example multiple `DoctrineGateway` are displayed when test cases are requested, even though they come from different classes (see ctrl+shift+T shortcut on implementation class).

I'm unsure what to do with duplicated code detected by sonar cloud - it seems to come from a fixture file, which to be honest I don't see duplicated anywhere in terms of data contained :confused:

adamwojs commented:
> +1 but please create/update integration test. It's PAPI change

@Steveb-p I meant `\eZ\Publish\API\Repository\Tests\URLServiceTest` here :wink:

Steveb-p commented:
> > +1 but please create/update integration test. It's PAPI change
> 
> @Steveb-p I meant `\eZ\Publish\API\Repository\Tests\URLServiceTest` here

This PAPI is unnaffected.

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Tests/URL/Gateway/DoctrineDatabaseTest.php:
```diff
@@ -0,0 +1,115 @@
+     *
+     * @var \eZ\Publish\Core\Persistence\Legacy\URL\Gateway\DoctrineDatabase
+     */
+    protected $gateway;
```
```suggestion
    private $gateway;
```

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Tests/URL/Gateway/DoctrineDatabaseTest.php:
```diff
@@ -0,0 +1,115 @@
+    /**
+     * @var array[]
+     */
+    protected $fixtureData;
```
```suggestion
    private $fixtureData;
```

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Tests/URL/Gateway/DoctrineDatabaseTest.php:
```diff
@@ -0,0 +1,115 @@
+     */
+    protected $gateway;
+
+    /**
+     * @var array[]
+     */
```
```suggestion
    /** @var array[] */
```

adamwojs commented on eZ/Publish/API/Repository/Values/URL/URLQuery.php:
```diff
@@ -47,4 +47,11 @@ class URLQuery extends ValueObject
+     *
+     * @var bool
+     */
+    public $performCount = true;
```
I guess `\eZ\Publish\API\Repository\Values\URL\SearchResult::$totalCount` docblock should be updated as well to `int|null`

Steveb-p replied:
Changed in caf54feb.

I didn't change the default value. Technically it's our internal VO and there is little chance that anyone creates those on their own, but for the sake of BC I've left it as is.

In case of our code we always pass the proper value in `totalCount` field in constructor, even if it's `null`.

tomaszszopinski reviewed (approved):
QA approved on Ibexa Exp 3.2.7-dev.

Steveb-p commented:
Merged up in 

* `1.3` - 68bccd7a
* `master` - 5a5d8dde
